### ISSUE TITLE
LIME Text explanation with Keras (using embedding) is not working 

### ISSUE THREAD

**Issue by giuliaass** [rstudio/keras#1174]
Hi,

I'm working on a Text classification Problem with Keras and lime.
My code talks about credit risk which is a classification problem, Yes or No,
When I predict the outcome for the first four Test Data I get: Yes No No Yes.
But when I explain my model with lime I get for the first four cases: No Yes Yes Yes.
Why is this the case and how can I fix it?

Thank you in advance!!

My Code is the following:

```
y_pred <- model_CNN_K_2T %>% predict_classes(x_test)
y_pred_prob <- model_CNN_K_2T %>% predict(x_test)

########################

Explanation of the model

########################

#install.packages("lime")
library(lime)

class(model_CNN_K_2T)

Setup lime::model_type() function for keras

model_type.keras.engine.sequential.Sequential <- function(x, ...) {
return("classification")
}

Setup lime::predict_model() function for keras

predict_model.keras.engine.sequential.Sequential <- function(x, newdata, type, ...) {
pred <- predict_proba(object = x, x = as.matrix(newdata))
return(data.frame(Yes = pred, No = 1 - pred))
}

predict_model.keras.engine.sequential.Sequential(x = model_CNN_K_2T, newdata = x_test, type = 'raw') %>% tibble::as_tibble()

get_embedding_explanation <- function(text) {
tokenizer %>% fit_text_tokenizer(text)
text_to_seq <- texts_to_sequences(tokenizer, text)
sentences <- text_to_seq %>% pad_sequences(maxlen = maxlen)
}

sentence_to_explain <- as.character(Refinitiv_data_test$Text[1:4])

explainer <- lime(Refinitiv_data_train$Text, model = model_CNN_K_2T, preprocess = get_embedding_explanation, bin_continuous = FALSE)

Get explanation for the first 20 words

explanation <- lime::explain(sentence_to_explain, explainer, n_features = 20, n_labels = 1, kernel_width = 0.5)

Different graphical ways to show the same information

plot_text_explanations(explanation)
plot_features(explanation)

Show Prediction

y_pred[1:4]
y_pred_prob[1:4]
```

**Comment by dfalbel** [comment]
Can you file a reproducible example? I can't reproduce the problem because I don't know what `model_CNN_K_2T` is for instance.

**Comment by giuliaass** [comment]
Hi, thank you a lot for your answer!
Can I send you the data and the code by mail?

**Comment by dfalbel** [comment]
I wonder if it's possible to reproduce the problem with easily generated data eg, with `runif` or other. THe model doesn't need to work good... it's just to be able to reproduce the error.

**Comment by giuliaass** [comment]
I tried to reproduced my problem with a simple example: A binary classification problem.
I divided the data in train and test data. And made predictions for the test data.
When I use lime to explain the first three sentences of the test data lime predicts for me the following labels:
Case 1: 1 (--> 0)
Case 2: 1 (--> 0)
Case 3: 2 (--> 1)

But my Keras model predicts Case 1: 1; Case 2: 0; Case 3: 1

Why is this the case or is there something wrong with my model?

`#' This example shows how to use lime to explain text data.

library(readr)
library(dplyr)
library(keras)
library(tidyverse)

# Download and unzip data

activity_url <- "https://archive.ics.uci.edu/ml/machine-learning-databases/00461/drugLib_raw.zip"
temp <- tempfile()
download.file(activity_url, temp)
unzip(temp, "drugLibTest_raw.tsv")

# Read dataset
df <- read_delim('drugLibTest_raw.tsv',delim = '\t')
unlink(temp)

# Select only rating and text from the whole dataset
df = df %>% select(rating,commentsReview) %>% mutate(rating = if_else(rating >= 8, 0, 1))

# Train and test data
set.seed(123)
train_indices<- sample(1:nrow(df), 0.7 * nrow(df))
test_indices <- setdiff(1:nrow(df), train_indices)

df_train <- df[train_indices, ]
df_test <- df[test_indices, ]

y_train <- df_train$rating
y_test <- df_test$rating

# text_tokenizer helps us to turn each word into integers. By selecting maximum number of features
# we also keep the most frequent words. Additionally, by default, all punctuation is removed.

max_features <- 1000
tokenizer <- text_tokenizer(num_words = max_features)

# Then, we need to fit the tokenizer object to our text data
text <- as.character(df_train$commentsReview)
tokenizer %>% fit_text_tokenizer(text)

# Finally, we can replace words in dataset with integers
text_seqs <- texts_to_sequences(tokenizer, text)

text_seqs %>% head(3)

# Define the parameters of the keras model

maxlen <- 15
batch_size <- 32
embedding_dims <- 50
filters <- 64
kernel_size <- 3
hidden_dims <- 50
epochs <- 15

# As a final step, restrict the maximum length of all sequences and create a matrix as input for model
x_train <- text_seqs %>% pad_sequences(maxlen = maxlen)

# Create a model
model <- keras_model_sequential() %>% 
  layer_embedding(max_features, embedding_dims, input_length = maxlen) %>%
  layer_dropout(0.2) %>%
  layer_conv_1d(
    filters, kernel_size, 
    padding = "valid", activation = "relu", strides = 1
  ) %>%
  layer_global_max_pooling_1d() %>%
  layer_dense(hidden_dims) %>%
  layer_dropout(0.2) %>%
  layer_activation("relu") %>%
  layer_dense(1) %>%
  layer_activation("sigmoid")

# Compile
model %>% compile(
  loss = "binary_crossentropy",
  optimizer = "adam",
  metrics = "accuracy"
)

# Run
hist <- model %>%
  fit(
    x_train,
    y_train,
    batch_size = batch_size,
    epochs = epochs,
    validation_split = 0.1
  )

text1 <- as.character(df_test$commentsReview)
tokenizer1 <- text_tokenizer(num_words = max_features) %>% fit_text_tokenizer(text1)
sequences1 <- texts_to_sequences(tokenizer1, text1)

x_test <- pad_sequences(sequences1, maxlen = maxlen)

y_pred <- model %>% predict_classes(x_test)

# Understanding lime for Keras Embedding Layers

# In order to explain a text with LIME, we should write a preprocess function
# which will help to turn words into integers. Therefore, above mentioned steps 
# (how to encode a text) should be repeated BUT within a function. 
# As we already have had a tokenizer object, we can apply the same object to train/test or a new text.

get_embedding_explanation <- function(text) {
  
  tokenizer %>% fit_text_tokenizer(text)
  
  text_to_seq <- texts_to_sequences(tokenizer, text)
  sentences <- text_to_seq %>% pad_sequences(maxlen = maxlen)
}

library(lime)

# Lets choose some text (3 rows) to explain
sentence_to_explain <- df_test$commentsReview[1:3]
sentence_to_explain
# You could notice that our input is just a plain text. Unlike tabular data, lime function 
# for text classification requires a preprocess fuction. Because it will help to convert a text to integers 
# with provided function. 
explainer <- lime(df_train$commentsReview, model = model, preprocess = get_embedding_explanation)

# Get explanation for the first 10 words
explanation <- explain(sentence_to_explain, explainer, n_labels = 1, n_features = 10,n_permutations = 1e4)

# Different graphical ways to show the same information

plot_text_explanations(explanation)
plot_features(explanation)
y_pred[1:3]

`

**Comment by giuliaass** [comment]
Hi,

I tried to reproduced my problem with a simple example: A binary classification problem.
I divided the data in train and test data. And made predictions for the test data.
When I use lime to explain the first three sentences of the test data lime predicts for me the following labels:
Case 1: 1 (--> 0)
Case 2: 1 (--> 0)
Case 3: 2 (--> 1)

But my Keras model predicts Case 1: 1; Case 2: 0; Case 3: 1

Why is this the case or is there something wrong with my model?

Thank you !


`library(readr)
library(dplyr)
library(keras)
library(tidyverse)


activity_url <- "https://archive.ics.uci.edu/ml/machine-learning-databases/00461/drugLib_raw.zip"
temp <- tempfile()
download.file(activity_url, temp)
unzip(temp, "drugLibTest_raw.tsv")

df <- read_delim('drugLibTest_raw.tsv',delim = '\t')
unlink(temp)

df = df %>% select(rating,commentsReview) %>% mutate(rating = if_else(rating >= 8, 0, 1))

set.seed(123)
train_indices<- sample(1:nrow(df), 0.7 * nrow(df))
test_indices <- setdiff(1:nrow(df), train_indices)

df_train <- df[train_indices, ]
df_test <- df[test_indices, ]

y_train <- df_train$rating
y_test <- df_test$rating

max_features <- 1000
tokenizer <- text_tokenizer(num_words = max_features)

text <- as.character(df_train$commentsReview)
tokenizer %>% fit_text_tokenizer(text)

text_seqs <- texts_to_sequences(tokenizer, text)

text_seqs %>% head(3)


maxlen <- 15
batch_size <- 32
embedding_dims <- 50
filters <- 64
kernel_size <- 3
hidden_dims <- 50
epochs <- 15

x_train <- text_seqs %>% pad_sequences(maxlen = maxlen)

model <- keras_model_sequential() %>% 
  layer_embedding(max_features, embedding_dims, input_length = maxlen) %>%
  layer_dropout(0.2) %>%
  layer_conv_1d(
    filters, kernel_size, 
    padding = "valid", activation = "relu", strides = 1
  ) %>%
  layer_global_max_pooling_1d() %>%
  layer_dense(hidden_dims) %>%
  layer_dropout(0.2) %>%
  layer_activation("relu") %>%
  layer_dense(1) %>%
  layer_activation("sigmoid")

model %>% compile(
  loss = "binary_crossentropy",
  optimizer = "adam",
  metrics = "accuracy"
)

hist <- model %>%
  fit(
    x_train,
    y_train,
    batch_size = batch_size,
    epochs = epochs,
    validation_split = 0.1
  )

text1 <- as.character(df_test$commentsReview)
tokenizer1 <- text_tokenizer(num_words = max_features) %>% fit_text_tokenizer(text1)
sequences1 <- texts_to_sequences(tokenizer1, text1)

x_test <- pad_sequences(sequences1, maxlen = maxlen)

y_pred <- model %>% predict_classes(x_test)


get_embedding_explanation <- function(text) {
  
  tokenizer %>% fit_text_tokenizer(text)
  
  text_to_seq <- texts_to_sequences(tokenizer, text)
  sentences <- text_to_seq %>% pad_sequences(maxlen = maxlen)
}

library(lime)

sentence_to_explain <- df_test$commentsReview[1:3]
sentence_to_explain

explainer <- lime(df_train$commentsReview, model = model, preprocess = get_embedding_explanation)


explanation <- explain(sentence_to_explain, explainer, n_labels = 1, n_features = 10,n_permutations = 1e4)


plot_text_explanations(explanation)
plot_features(explanation)
y_pred[1:3]

`

**Comment by giuliaass** [comment]
Hi,

I tried to reproduced my problem with a simple example: A binary classification problem.
I divided the data in train and test data. And made predictions for the test data.
When I use lime to explain the first three sentences of the test data lime predicts for me the following labels: 1 0 1 1 0 1

But my Keras model predicts 1 0 1 1 1 1

Why is this the case or is there something wrong with my model?

Thank you !

library(readr)
library(dplyr)
library(keras)
library(tidyverse)


activity_url <- "https://archive.ics.uci.edu/ml/machine-learning-databases/00461/drugLib_raw.zip"
temp <- tempfile()
download.file(activity_url, temp)
unzip(temp, "drugLibTest_raw.tsv")

df <- read_delim('drugLibTest_raw.tsv',delim = '\t')
unlink(temp)

df = df %>% select(rating,commentsReview) %>% mutate(rating = if_else(rating >= 8, 0, 1))

set.seed(123)
train_indices<- sample(1:nrow(df), 0.7 * nrow(df))
test_indices <- setdiff(1:nrow(df), train_indices)

df_train <- df[train_indices, ]
df_test <- df[test_indices, ]

y_train <- df_train$rating
y_test <- df_test$rating

max_features <- 1000
tokenizer <- text_tokenizer(num_words = max_features)

text <- as.character(df_train$commentsReview)
tokenizer %>% fit_text_tokenizer(text)

text_seqs <- texts_to_sequences(tokenizer, text)

text_seqs %>% head(3)


maxlen <- 15
batch_size <- 32
embedding_dims <- 50
filters <- 64
kernel_size <- 3
hidden_dims <- 50
epochs <- 15

x_train <- text_seqs %>% pad_sequences(maxlen = maxlen)

model <- keras_model_sequential() %>% 
  layer_embedding(max_features, embedding_dims, input_length = maxlen) %>%
  layer_dropout(0.2) %>%
  layer_conv_1d(
    filters, kernel_size, 
    padding = "valid", activation = "relu", strides = 1
  ) %>%
  layer_global_max_pooling_1d() %>%
  layer_dense(hidden_dims) %>%
  layer_dropout(0.2) %>%
  layer_activation("relu") %>%
  layer_dense(1) %>%
  layer_activation("sigmoid")

model %>% compile(
  loss = "binary_crossentropy",
  optimizer = "adam",
  metrics = "accuracy"
)

hist <- model %>%
  fit(
    x_train,
    y_train,
    batch_size = batch_size,
    epochs = epochs,
    validation_split = 0.1
  )

text1 <- as.character(df_test$commentsReview)
tokenizer1 <- text_tokenizer(num_words = max_features) %>% fit_text_tokenizer(text1)
sequences1 <- texts_to_sequences(tokenizer1, text1)

x_test <- pad_sequences(sequences1, maxlen = maxlen)

y_pred <- model %>% predict_classes(x_test)
y_pred_prob <- model %>% predict(x_test)


get_embedding_explanation <- function(text) {
  
  tokenizer %>% fit_text_tokenizer(text)
  
  text_to_seq <- texts_to_sequences(tokenizer, text)
  sentences <- text_to_seq %>% pad_sequences(maxlen = maxlen)
}

library(lime)

class(model)

# Setup lime::model_type() function for keras
model_type.keras.engine.sequential.Sequential <- function(x, ...) {
  return("classification")
}

# Setup lime::predict_model() function for keras
predict_model.keras.engine.sequential.Sequential <- function(x, newdata, type, ...) {
  pred <- predict_proba(object = x, x = as.matrix(newdata))
  return(data.frame("1" = pred, "0" = 1 - pred))
}

predict_model.keras.engine.sequential.Sequential(x = model, newdata = x_test, type = 'raw') %>% tibble::as_tibble()

sentence_to_explain <- df_test$commentsReview[1:6]
sentence_to_explain

explainer <- lime(df_train$commentsReview, model = model, preprocess = get_embedding_explanation)

explanation <- explain(sentence_to_explain, explainer, n_labels = 1, n_features = 10,n_permutations = 1e4)

plot_features(explanation)

y_pred[1:6]
y_pred_prob[1:6]

**Comment by dfalbel** [comment]
@giuliaass I think the issue in your code is that you are using different tokenizers for training and prediction ad they should be the same.

The tokenizer in Keras maps words to integer tokens eg. 'the'-> 1, 'other'-> 2  and etc. If a different mapping is used for prediction the results will be wrong.

I have modified your code to not `fit_tokenizer` again and just use the tokenizer you use for training and everything looks fine, including the predictions from lime are identical to keras.

**Comment by giuliaass** [comment]
@dfalbel Thanks for the comment! I have now changed this so that the same tokenizer is used.

Test data:
text1 <- as.character(df_test$commentsReview)
sequences1 <- texts_to_sequences(tokenizer, text1)
x_test <- pad_sequences(sequences1, maxlen = maxlen)

Preprocess function:
get_embedding_explanation <- function(text) {
  text_to_seq <- texts_to_sequences(tokenizer, text)
  sentences <- text_to_seq %>% pad_sequences(maxlen = maxlen)

Unfortunately, that still doesn't work for me and I continue to have the same problem. Can you maybe share the code that worked for you?
Thank you!

**Comment by dfalbel** [comment]
Hi @giuliaass ,

I did the same changes as you did. I think the problem here is that the linear models fitted by lime is resulting in not resulting in a good approximation of the recurrent neural network.

You can see the 'Explanation Fit' field in this plot:

![image](https://user-images.githubusercontent.com/4706822/108384807-d7657000-71e9-11eb-9f78-a699aca6e8b7.png)

I suggest this reading this for ideas on how to make sure that lime is getting good local approximations: https://uc-r.github.io/lime
